c7n/resources/ec2.py

```diff
@@ -1698,3 +1698,55 @@ def get_instance_attribute(self, resources, attribute):
             keys.remove('InstanceId')
             resource['c7n:attribute-%s' % attribute] = fetched_attribute[
                 keys[0]]
+
+
+@resources.register('launch-template')
+class LaunchTemplate(query.QueryResourceManager):
+
+    class resource_type(object):
+        service = 'ec2'
+        type = 'LaunchTemplate'
+        id = 'LaunchTemplateId'
+        name = 'LaunchTemplateName'
+        date = 'CreateTime'
+        dimension = 'LaunchTemplateName'
+        enum_spec = (
+            'describe_launch_templates', 'LaunchTemplates', None)
+        filter_name = 'LaunchTemplateNames'
+        filter_type = 'list'
+        config_type = "AWS::EC2::LaunchTemplate"
+
+
+class LaunchTemplateDataFilterBase(object):
+
+    template_configs = {}
+
+    def initialize(self, asgs):
+        super(LaunchTemplateDataFilterBase, self).initialize(asgs)
+        config_names = set()
+        skip = []
+
+        for a in asgs:
+            if 'LaunchTemplate' not in a and 'LaunchConfigurationName' not in a:
```

Comment:
the second part of the this conditional looks odd, why would the template filter be concerned about a launch config?

Reply:
As far as I know, the unecrypted filter inherits from both LaunchConfigFilterBase and LaunchTemplateDataFilterBase and initialization is ran on both through one set of ASGs which is modified during initialization. In order to make sure that asgs are not incorrectly removed from the set prematurely, we must check both conditions.

Reply:
That makes me feel a little suspect about how this is being wired together, its not clean on composition, if every implementation behind the interface has to be aware of all the other implementations. I'll try and spend some more time with this pr tomorrow.

Reply:
Any new thoughts about the implementation?